### PR TITLE
Add a vertical toolbar for zoom out mode

### LINKED ISSUE #59737: Explore a vertical toolbar for zoom out mode
I'd like to explore bringing back a less intrusive vertical toolbar for when zoom-out mode is engaged. This way, the block toolbar does obscure the patterns above the selected pattern, especially as the patterns are visually smaller when they are zoomed out. 

Dependent on https://github.com/WordPress/gutenberg/issues/60124.

To start, let's expose the block movers, the shuffle control (https://github.com/WordPress/gutenberg/issues/45552), and a delete control. 

### Visual 

CleanShot 2024-03-18 at 13 55 40

Related #56808, #56888, #56886, #56894.

### PR DESCRIPTION
**Pull request by MaggieCabrera**
## What?

Closes https://github.com/WordPress/gutenberg/issues/59737

Ariakit supports vertical toolbars so we implemented a new variant to the main `Toolbar` component and applied it to a new toolbar specifically for the zoom-out mode.

## Why?

Zoom-out mode needs a toolbar that doesn't cover the content we are inserting. We needed to leverage the Ariakit toolbar for accessibility, so the black pseudo-toolbar that we were using wasn't enough.

## How?

By implementing a toolbar that extends `NavigableToolbar` and adding a new orientation prop that ariakit uses

## Testing Instructions

Open the inserter and open a pattern category
Click on a section, the toolbar should be visible on the side. All buttons should work as usual

### Testing Instructions for Keyboard

When you focus on a section, move focus to the toolbar and use up and down arrows to navigate between the toolbar buttons

## Screenshots or screencast 

Screenshot 2024-05-10 at 11 38 44

### CODE REVIEW

**Comment by MaggieCabrera** [comment]
I'm going to focus first on the other items in https://github.com/WordPress/gutenberg/issues/60124 before continuing this exploration

**Comment by scruffian** [comment]
I think it would be nice if the toolbar stuck to the top of the screen.

File: packages/components/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 ### Internal
 
 -   Replaced `classnames` package with the faster and smaller `clsx` package ([#61138](https://github.com/WordPress/gutenberg/pull/61138)).
+-   Add a vertical toolbar for zoom out mode ([#60123](https://github.com/WordPress/gutenberg/pull/60123))
```
Comment (mirka) [inline packages/components/CHANGELOG.md]:
```suggestion
-   `Toolbar`: Add support for `vertical` orientation ([#60123](https://github.com/WordPress/gutenberg/pull/60123)).
```

The changelog is for all consumers of the wp-components package, so let's phrase this in a way that's relevant to that audience. Also, I think this should be considered an Enhancement, rather than just Internal.

File: packages/components/src/toolbar/toolbar/style.scss [hunk 2]
```diff
@@ -13,9 +13,15 @@
 
 		& > .components-toolbar-group {
 			border-right: none;
+			background: transparent;
```
Comment (mirka) [inline packages/components/src/toolbar/toolbar/style.scss]: This background color change could be a problem for existing usages of the `unstyled` variant. There are a couple of safer ways to achieve this from the consumer, but I think the simplest way for now would be to pass your own CSS class to the `ToolbarGroup` instance(s) you want transparent.

```
.my-toolbar-group {
  background: transparent;
}

<ToolbarGroup className="my-toolbar-group" />
```

File: packages/components/src/toolbar/toolbar/types.ts [hunk 3]
```diff
@@ -22,4 +22,11 @@ export type ToolbarProps = {
 	 * @default undefined
 	 */
 	variant?: 'unstyled' | undefined;
+	/**
+	 * Orientation of the toolbar keyboard interactions.
+	 * horizontal or vertical
```
Comment (mirka) [inline packages/components/src/toolbar/toolbar/types.ts]:
```suggestion
```

I think we can omit this because the docgen will show the available enums from the type data.

File: packages/components/src/toolbar/toolbar/types.ts [hunk 4]
```diff
@@ -22,4 +22,11 @@ export type ToolbarProps = {
...
+	 *
+	 * @default 'horizontal'
+	 */
+	orientation?: 'horizontal' | 'vertical';
```
Comment (mirka) [inline packages/components/src/toolbar/toolbar/types.ts:30]: Let's document this prop in the component README as well.

**Review by mirka — COMMENTED** [review]
Aside from some small tweaks I suggested, this looks good from the components side 👍

**Comment by MaggieCabrera** [comment]
I addressed the review of the component's changes. Do we think this is a good starting point for the toolbar on zoom out mode? /cc @richtabor

**Comment by MaggieCabrera** [comment]
> I think it would be nice if the toolbar stuck to the top of the screen.

like a fixed toolbar? I'm not sure if I understand what this would entail

**Comment by scruffian** [comment]
> I think it would be nice if the toolbar stuck to the top of the screen.


~This is what happens for a standard toolbar:~
Screenshot 2024-06-18 at 12 32 26

~Can we just add position: sticky?~

Edit: this is exactly what happens in this PR now!

**Comment by scruffian** [comment]
I rebased this and now I'm seeing a few visual issues, sorry. I'm not sure if these are related to the rebase:

Screenshot 2024-06-18 at 12 40 23
Screenshot 2024-06-18 at 12 39 57

File: packages/block-editor/src/components/block-tools/zoom-out-toolbar.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+		isBlockTemplatePart,
+		isNextBlockTemplatePart,
+		isPrevBlockTemplatePart,
+		canRemove,
+		canMove,
+	} = selected;
+
+	const { removeBlock } = useDispatch( blockEditorStore );
+
+	const classNames = clsx( 'zoom-out-toolbar', {
```
Comment (scruffian) [inline packages/block-editor/src/components/block-tools/zoom-out-toolbar.js:82]: I wonder if `vertical-toolbar` is a better name as this could be used in other context.

File: packages/components/src/toolbar/toolbar-group/style.scss [hunk 6]
```diff
@@ -14,6 +13,10 @@
 		margin: 0;
 	}
 
+	.components-accessible-toolbar:not(.zoom-out-toolbar) & {
+		background: $white;
```
Comment (scruffian) [inline packages/components/src/toolbar/toolbar-group/style.scss]: I think it would be better to just remove this altogether, but there is a risk that it's used for something...

File: packages/editor/src/components/visual-editor/index.js [hunk 7]
```diff
@@ -341,7 +341,7 @@ function VisualEditor( {
 	const zoomOutProps = isZoomOutMode
 		? {
 				scale: 'default',
-				frameSize: '20px',
+				frameSize: '48px',
```
Comment (scruffian) [inline packages/editor/src/components/visual-editor/index.js:344]: I made the frame size larger so that at small resolutions there is still space for the vertical toolbar: 
Screenshot 2024-06-18 at 14 57 53

File: packages/components/src/toolbar/toolbar/style.scss [hunk 8]
```diff
@@ -20,6 +20,12 @@
 
 .components-accessible-toolbar,
 .components-toolbar {
+
+	&[aria-orientation="vertical"] {
```
Comment (scruffian) [inline packages/components/src/toolbar/toolbar/style.scss:24]: I moved this down so that it also targets `components-toolbar` as well as `components-accessible-toolbar`

**Comment by scruffian** [comment]
I have fixed the issues above.

Sometimes the toolbar ends up on the right, which is a bit surprising but it does work ok:
Screenshot 2024-06-18 at 23 20 16

**Review by scruffian — COMMENTED** [review]
This is looking pretty good to me. My main concerns are:
1. Should we rename zoom out toolbar to vertical toolbar or something more generic?
2. Is there too much duplication in the `ZoomOutPopover` and `ZoomOutToolbar` components - would it be better to compose the base components?

**Comment by MaggieCabrera** [comment]
> This is looking pretty good to me. My main concerns are:
> 
>     1. Should we rename zoom out toolbar to vertical toolbar or something more generic?
> 

I think the fact that is vertical is just an extra. The CSS for the component will be available to any other future vertical toolbars, the rest of the changes are specific to the needs of zoom out (mainly the contents of the toolbar we need it to have in this specific instance of it)

**Review by scruffian — APPROVED** [review]
LGTM, if you're happy with my changes.

File: packages/components/src/toolbar/toolbar/style.scss [hunk 9]
```diff
@@ -20,12 +20,6 @@
 
 .components-accessible-toolbar,
 .components-toolbar {
-
-	&[aria-orientation="vertical"] {
-		flex-direction: column;
-		align-items: center;
-	}
-
```
Comment (MaggieCabrera) [inline packages/components/src/toolbar/toolbar/style.scss]: Why did you move this @scruffian? I put it here so the toolbar component can be used vertically in any other setting, not just zoom out. Since we added the prop for it too, without the CSS there is no visual change if orientation is set to vertical.

**Comment by MaggieCabrera** [comment]
> I have fixed the issues above.
> 
> Sometimes the toolbar ends up on the right, which is a bit surprising but it does work ok: Screenshot 2024-06-18 at 23 20 16

I think https://github.com/WordPress/gutenberg/pull/60123/commits/6d5ad628ca362723e6b1e829cef792804821bf01 fixes that

**Comment by richtabor** [comment]
I noticed that sometimes it's collapsing a bit (safari):

https://github.com/WordPress/gutenberg/assets/1813435/0f502eae-f1cd-4563-9a35-25b28e1e58b5

**Comment by richtabor** [comment]
Can we also remove this double focus style (the white): 
CleanShot 2024-06-25 at 09 54 34

**Comment by MaggieCabrera** [comment]
Testing this PR with @ajlende and @jeryj we have found that showing the toolbar like this highlights a bunch of issues that the zoom out mode is suffering from:

- On resize, the calculation of the position of the toolbar (`BlockPopover`) seems inconsistent and jumpy. When we use the profiler we can see thousands of calls to `mouseleave`. We see this on trunk too, but this PR makes it very apparent.
- In my own testing I see the whole canvas and the toolbar and inserter jump when the scrollbar appears/disappears. This is not created by this PR, it's just more noticeable here.

I have seen the padding change that Rich mentions once but I've not been able to reproduce again. I'm looking at it.

File: packages/block-editor/src/components/navigable-toolbar/index.js [hunk 10]
```diff
@@ -238,12 +238,7 @@ export default function NavigableToolbar( {
 	}
 
 	return (
-		<NavigableMenu
-			orientation="horizontal"
```
Comment (MaggieCabrera) [inline packages/block-editor/src/components/navigable-toolbar/index.js:242]: I think we don't need that because orientation is already being passed inside `props`, I might be missing something here @jeryj

Comment (jeryj) [reply packages/block-editor/src/components/navigable-toolbar/index.js:242]: I think we did remove it. It looks like the updated version is `role="toolbar" ref={ toolbarRef } { ...props }`. Maybe we should pass a default of `orientation: horizontal` for backwards compatibility though? In case they don't pass an orientation, it would no longer be there for older versions.

Comment (jeryj) [reply packages/block-editor/src/components/navigable-toolbar/index.js:242]: Oh, actually we should leave it there. `<NavigableMenu />` has a default orientation of `vertical`. If we remove this, then anyone using this component that's not passing an orientation will have their toolbar orientation switch from `horizontal` to `vertical`.

Comment (MaggieCabrera) [reply packages/block-editor/src/components/navigable-toolbar/index.js:242]: Good catch, I'll revert that change then

File: packages/block-editor/src/components/block-tools/style.scss [hunk 11]
```diff
@@ -270,3 +270,16 @@
...
+
+	.components-button:focus::before {
+		box-shadow: 0 0 0 var(--wp-admin-border-width-focus) var(--wp-admin-theme-color);
+	}
+
+	.components-toolbar,
+	.components-toolbar-group {
+		background: none;
+		border: none;
+	}
```
Comment (ciampo) [inline packages/block-editor/src/components/block-tools/style.scss]: 👋  Hi there, and sorry for just casually popping by while this PR has been open for quite some time already!

Just wanted to flag that `.components-*` class names are intended to be private implementation details, and that style overrides can lead to unexpected breakages if the component updates its internals.

Can we avoid using those style overrides? Are those changes something that could be applied directly at the component level?

Comment (MaggieCabrera) [reply packages/block-editor/src/components/block-tools/style.scss]: well, ideally I think these changes should go on the component, since here we are using the `unstyled` version of the toolbar, these backgrounds/border/shadow are being applied by the original component and in my opinion they shouldn't when the `unstyled` attr is present. @mirka advised [against](https://github.com/WordPress/gutenberg/pull/60123#discussion_r1603589830) the initial change, so when we found the other problems, we decided to add our own class to the toolbar instead

Comment (MaggieCabrera) [reply packages/block-editor/src/components/block-tools/style.scss:284]: I think we can go without some of this CSS, but I'm unsure about the focus, I'll get back to you

Comment (ciampo) [reply packages/block-editor/src/components/block-tools/style.scss]: I see! Thank you for the context, I missed it initially.

A good improvement, as also suggested by Lena in the comment you link, could still be to pass custom classnames to the component. That way, you can use those custom classnames (instead of private `.components-*` classnames) to apply the style overrides. Not ideal, but an improvement nonetheless. Would you be able to do that?

Thank you!

Comment (ciampo) [reply packages/block-editor/src/components/block-tools/style.scss]: Just seen https://github.com/WordPress/gutenberg/pull/60123/commits/b690f1527b580e09245e899eac3bd9297d5ef6de — thank you again!

Comment (MaggieCabrera) [reply packages/block-editor/src/components/block-tools/style.scss]: I still think it would be a good idea to revisit if we should make this change in the component instead. But that can happen on a different PR. Thanks for the review! If you don't have any other issues with the component changes on this one, I will merge it soon

Comment (ciampo) [reply packages/block-editor/src/components/block-tools/style.scss]: Agreed — and nope, no further comments from my side 🚀
